include/swift/AST/TypeCheckRequests.h

```diff
@@ -2020,6 +2025,35 @@ class PatternTypeRequest
   }
 };
 
+/// Type-checks a `@differentiable` attribute and returns the resolved parameter
+/// indices on success. On failure, emits diagnostics and returns `nullptr`.
+///
+/// Currently, this request resolves other `@differentiable` attribute
+/// components but mutates them in place:
+/// - `JVPFunction`
+/// - `VJPFunction`
+/// - `DerivativeGenericSignature`
+class DifferentiableAttributeTypeCheckRequest
+    : public SimpleRequest<DifferentiableAttributeTypeCheckRequest,
+                           IndexSubset *(DifferentiableAttr *),
```

Comment:
Currently, `DifferentiableAttributeTypeCheckRequest` takes a `DifferentiableAttr *`and returns an `IndexSubset *` representing parameter indices. It also resolves two `FuncDecl *` (`JVPFunction` and `VJPFunction`) and a `GenericSignature` (`DerivativeGenericSignature`), mutating them in-place in the `DifferentiableAttr *`.

This works fine for now. I thought of two ways to refactor the request to avoid mutation:
* Make `DifferentiableAttributeTypeCheckRequest` return a tuple/struct of the resolved components.
  * This caused many request cycles, since `DifferentiableAttr::get{JVPFunction,VJPFunction,DerivativeGenericSignature}` now all trigger the request. I haven't debugged further.
* Make an individual request for resolving each `@differentiable` attribute component: original `AbstractFuncDecl`(s), parameter indices, JVP/VJP `FuncDecl` (if specified), derivative `GenericSignature` (if specified).
  * I haven't tried this. It seems like a fair amount of work, and the upside is unclear.

Reply:
History: https://github.com/apple/swift/pull/28017 moved `@differentiable` attribute type-checking logic from `AttributeChecker::visitDifferentiableAttr` to `DifferentiableAttributeTypeCheckRequest::evaluate`.

The current request approach (return `IndexSubset *` but mutate other components in-place) is the first one that worked.